### ISSUE TITLE
sd3.5 training script

### ISSUE THREAD

**Issue by SwayStar123** [caitree/scfm#2]
Will it be possible for you to include the sd3.5 training script? 

I think it will be more relevant to the broader research community as it will be usable for non guidance distilled models, whereas the flux trainer is only for guidance distilled

**Comment by SwayStar123** [comment]
I tried modifying the code to work with non-guidance distilled models, unforunately am unable to replicate the results

400 train steps, there is significant noise/perturbation added to the generations. Output is worse than just using the teacher at 8 steps normally
<img width="2048" height="1166" alt="Image" src="https://github.com/user-attachments/assets/8913aafe-c38c-492b-b025-988dcfd2785d" />
800 train steps, perturbations are worsening, artifacts are added
<img width="2048" height="1166" alt="Image" src="https://github.com/user-attachments/assets/5980156e-59dd-4ea0-bfe0-1d12112c56df" />


I am trying to distill the bria model, which is similar to sd3.5 for the most part. 

@caitree

**Comment by SwayStar123** [comment]
Also i notice the non_shortcut_ratio is set to 0 in the bash script. This should be 0.6 right? Are there any other hyperparameters which are incorrect here?

**Comment by caitree** [comment]
Hi @SwayStar123 

Thank you very much for trying our method so quickly!

Apologies for the inconvenience — at the moment, I only have the resources and time to test the FLUX training script before releasing it, since I lost my previous environments and need to start from scratch.

If you’d like to give it a quick try, here is an [SD35L 8 step lora ckpt](https://huggingface.co/cxxai/scfm/blob/main/sd35l-scfm-lora-8-step.safetensors).  Using the prompt "_surreal photography in mid-air, Milky Way in the background, field of cosmic dust and glowing nebulas, an astronaut riding a majestic horse through the stars, weightless motion, shimmering starlight reflecting on the astronaut’s visor, cinematic lighting, ultra-detailed, ethereal atmosphere, dreamlike composition, long exposure astrophotography style, 8k resolution_" , with an _empty_ (null) negative prompt (note that other negative prompts are not recommended, since the checkpoint was only trained with an empty one), you should get the expected result (Left: 8-step teacher; Right: 8-step SCFM student.).

![Image](https://github.com/user-attachments/assets/4634abb7-e7a9-488a-93f8-0278f10a2172)

For re-creating this, just to clarify: have you applied **student velocity with CFG**? According to discussions around equation (16), all velocity calculations, including the student’s, should be CFG-rectified. That means the CFG part should still be applied after distillation, while other parts remain pretty much the same as in FLUX.  Specifically, for the student prediction, the code should look like:
```python
model_pred = transformer(
          hidden_states=torch.cat([noisy_model_input]*2),
          timestep=torch.cat([timesteps]*2)*1000,
          encoder_hidden_states=torch.cat([prompt_embeds, neg_prompt_embeds]),
          pooled_projections=torch.cat([pooled_prompt_embeds, neg_pooled_prompt_embeds]),
          return_dict=False,
    )[0]
model_pred, neg_model_pred = model_pred.chunk(2)
model_pred = model_pred + guidance * (model_pred - neg_model_pred)
```

Hope this clarifies, and let me know how it goes.

**Comment by caitree** [comment]
> Also i notice the non_shortcut_ratio is set to 0 in the bash script. This should be 0.6 right? Are there any other hyperparameters which are incorrect here?

FYI, _non_shortcut_ratio_ only controls the fraction of batch samples that use regular LoRA training versus distillation. For tasks that are purely distillation, it should be set to 0.  

Other hyperparameters looks good to me, perhaps you may consider loweirng _max_grad_norm_ ,  if you'd prefer not too high LoRA scales during inference.

P.S. Moderately restarting the EMA can help speed up training.

**Comment by SwayStar123** [comment]
Thank you for your response!

I managed to fix a few bugs and improve the quality alot. I managed to get good results with 100 train steps, however the velocity magnitudes slowly increase with train time and need to be normalized to obtain good results

original:
std deviation around 1.2, mean 0, min -6, max 5

distilled model at 100 steps:
std deviation is around 1.6, mean -0.3, min is -6.5, and max is 6 

distilled model at 200 steps:
std deviation around 3, mean 2.5, min -3.5, max 18

Unfortunately at 200 train steps, even normalization does not save the outputs, and the outputs are pure black.

The loss also slowly increases from ~0.05 at the start, to ~0.5-1 at 100 steps, and then increases alot to 30-100 at 200 steps

Do you have any ideas on what is the problem?


Without normalization at 100 steps:

<img width="2048" height="1166" alt="Image" src="https://github.com/user-attachments/assets/59f74caf-dbf0-4df7-a6d7-d9ebe22386d9" />

With normalization at 100 steps:

<img width="2048" height="1166" alt="Image" src="https://github.com/user-attachments/assets/efef9ed3-7313-4af7-920b-03c07a9ff925" />

200 steps with normalization (without normalization is also same, pure black):

<img width="2048" height="1166" alt="Image" src="https://github.com/user-attachments/assets/533ed558-26ca-40cf-8cc0-263cfa74c5ba" />

To answer your questions:

Yes I do CFG with both student, teacher, and cfg. I also implemented a guidance distillation version, which is also working pretty good, although it is accelerating the above turn to darkness (even at 100 steps some outputs are darker, but not full black)

**Comment by caitree** [comment]
Hi @SwayStar123 

Sorry for putting you through some trial and error, and I’m really glad to hear you’ve managed to get better results.

I haven’t used any normalization tricks for my previous runs — the loss barely increases during training (it usually stays around 0.0x).  (But it does seem that normalization indeed helps improve learning!)

Unfortunately, I don’t see any obvious issue that would cause your velocity magnitude and loss to grow at such a rapid pace :(.

I may need a few more days to investigate this further, and thanks for your patience and understanding.

**Comment by caitree** [comment]
Hi @SwayStar123 

Not sure if you’re still interested, but I just committed the training script for SD3.5. I gave it a quick test for 2000 training steps using the same prompt, and the results are as follows: ( Left: 8-step teacher; Mid: 32-step teacher; Right: 8-step SCFM student.).

![Image](https://github.com/user-attachments/assets/18e00b7e-3c69-40c8-bd56-86e685a34bdd)

It seems to be behaving well now.

Feel free to experiment with additional training tricks such as velocity normalization on top of it!

**Comment by SwayStar123** [comment]
The code is nearly identical to mine so not sure whats going wrong :/

I am using 8 gpus, thats the only difference i can find.

after 200 train steps my outputs turn into this

<img width="2048" height="1166" alt="Image" src="https://github.com/user-attachments/assets/be5e6583-94fd-4bb7-b71b-1d7c1da6b3e1" />

There might be some subtle bug i have introduced so i will try reproducing your results first and then trying to adapt it again

**Comment by SwayStar123** [comment]
Okay, using the code you have pushed, no changes to the code, using all the defaults, i get this at 100 steps training:

<img width="2048" height="1166" alt="Image" src="https://github.com/user-attachments/assets/853232fb-2171-4d12-bc2a-db9be2650e64" />

Is it because im using 8 gpus?

**Comment by SwayStar123** [comment]
I think the ema weights are not being synchronized, that may be the cause

**Comment by caitree** [comment]
Hi @SwayStar123 

Yes, exactly, I should’ve anticipated that issue. As I noted in the README, this version wasn’t built with multi-GPU support in mind. You’d probably need to add something like accelerator.broadcast or a shared buffer to make it work properly across devices.  Thanks for catching that.
